src/compas/geometry/transformations/quaternions.py

```diff
@@ -0,0 +1,324 @@
+
+    return abs(a - b) <= (atol + rtol * abs(b))
+
+# ----------------------------------------------------------------------
+
+
+def quaternion_norm(q):
+    """Calculates the length (euclidean norm) of a quaternion.
+
+    Parameters
+    ----------
+    q : list
+        Quaternion as a tuple of four real values [qw,qx,qy,qz].
+
+    Returns
+    -------
+    float
+        The length (euclidean norm) of a quaternion.
+
+    Note
+    ----
+    .. _mathworld quaternion norm
+    http://mathworld.wolfram.com/QuaternionNorm.html
+    """
+    return math.sqrt(sum([x*x for x in q]))
+
+
+def quaternion_unitize(q):
+    """Makes a quaternion unit-length.
+    Parameters
+    ----------
+    q : list
+        Quaternion as a tuple of four real values [qw,qx,qy,qz].
+
+    Returns
+    -------
+    list
+        A quaternion of length 1.0.
+    """
+    n = quaternion_norm(q)
+    if isclose(n, 0.0):
+        raise ValueError("The given quaternion has zero length.")
+    else:
+        return [x/n for x in q]
+
+
+def quaternion_is_unit(q):
+    """Checks if a quaternion is unit-length.
+    Parameters
+    ----------
+    q : list
+        Quaternion as a tuple of four real values [qw,qx,qy,qz].
+
+    Returns
+    -------
+    bool
+        True if the quaternion is unit-length, and False if otherwise.
+    """
+    n = quaternion_norm(q)
+    return isclose(n, 1.0)
+
+
+def quaternion_multiply(r, q):
+    """Multiplies two quaternions.
+    Parameters
+    ----------
+    r : list
+        Quaternion as a tuple of four real values [qw,qx,qy,qz].
+    q : list
+        Quaternion as a tuple of four real values [qw,qx,qy,qz].
+
+    Returns
+    -------
+    list
+        A quaternion p = r * q.
+
+    Note
+    ----
+    Multiplication of two quaternions r * q can be interpreted as applying rotation r to an orientation q,
+    provided both r and q are unit-length.
+    The result is also unit-length.
+    Multiplication of quaternions is not commutative!
+
+    .. _mathworld quaternion:
+    http://mathworld.wolfram.com/Quaternion.html
+    """
+    rw, rx, ry, rz = r
+    qw, qx, qy, qz = q
+    pw = rw*qw - rx*qx - ry*qy - rz*qz
+    px = rw*qx + rx*qw + ry*qz - rz*qy
+    py = rw*qy - rx*qz + ry*qw + rz*qx
+    pz = rw*qz + rx*qy - ry*qx + rz*qw
+    return [pw, px, py, pz]
+
+
+def quaternion_canonic(q):
+    """Converts a quaternion into a canonic form if needed.
+
+    Parameters
+    ----------
+    q : list
+        Quaternion as a tuple of four real values [qw,qx,qy,qz].
+
+    Returns
+    -------
+    list
+        A quaternion in a canonic form.
+
+    Note
+    ----
+    Canonic form means the scalar component is a non-negative number.
+
+    """
+    if q[0] < 0.0:
+        return [-x for x in q]
+    else:
+        return [x for x in q]
+
+
+def quaternion_conjugate(q):
+    """Conjugate of a quaternion.
+
+    Parameters
+    ----------
+    q : list
+        Quaternion as a tuple of four real values [qw,qx,qy,qz].
```

Comment:
type is indicated as `list`, but you describe it as a tuple.
please also mind the syntax: `[qw, qx, qy, qz]`

Reply:
and add code ticks around code: ``[qw, qx, qy, qz]``.
otherwise they are rendered as plain text.

Reply:
ok